lamindb/dev/db/_add.py

```diff
@@ -81,28 +81,84 @@ def add(  # type: ignore
             return results
         else:
             records = [model(**fields)]
-    for record in records:
-        if isinstance(record, DObject) and hasattr(record, "_local_filepath"):
-            upload_data_object(record, use_fsspec=use_fsspec)
+
     if session is None:  # assume global session
         session = settings.instance.session()
         settings.instance._cloud_sqlite_locker.lock()
         close = True
     else:
         close = False
+
+    added_records = []
+    raise_error = False
     for record in records:
-        session.add(record)
-    session.commit()
-    for record in records:
+        # commit metadata to database
+        prepare_filekey_metadata(record)
+        try:
+            session.add(record)
+            session.commit()
+            session.refresh(record)
```

Comment:
This is going to be called in line 117. If it triggers an exception here, it would also trigger an Exception in line 117.

Hence, it shouldn't be within a try-except.

Reply:
Consider dropping the `refresh()`, here. I don't see why you need it, but I may miss something.

Reply:
Implemented. This is no longer relevant with the single DB transaction implementation.